Sources_v3/StreamChat/Models/Attachment.swift

```diff
@@ -0,0 +1,310 @@
+//
+// Copyright © 2020 Stream.io Inc. All rights reserved.
+//
+
+import Foundation
+
+public typealias ChatMessageAttachment = _ChatMessageAttachment<DefaultExtraData.Attachment>
+
+@dynamicMemberLookup
+public struct _ChatMessageAttachment<ExtraData: AttachmentExtraData>: Hashable {
+    /// A title.
+    public let title: String
+    /// An author.
+    public let author: String?
+    /// A description text.
+    public let text: String?
+    /// A type (see `AttachmentType`).
+    public let type: AttachmentType
+    /// Actions from a command (see `Action`, `Command`).
+    public let actions: [AttachmentAction]
+    /// A URL.
+    public let url: URL?
+    /// An image preview URL.
+    public let imageURL: URL?
+    /// A file description (see `AttachmentFile`).
+    public let file: AttachmentFile?
+    /// An extra data for the attachment.
+    public let extraData: ExtraData
+    
+    // MARK: - Hashable
+    
+    var hash: String {
+        [title, author, text, type.rawValue, url?.absoluteString, imageURL?.absoluteString]
+            .compactMap { $0 }
+            .reduce("", +)
+    }
+    
+    public static func == (lhs: _ChatMessageAttachment<ExtraData>, rhs: _ChatMessageAttachment<ExtraData>) -> Bool {
+        lhs.title == rhs.title
+            && lhs.author == rhs.author
+            && lhs.text == rhs.text
+            && lhs.type == rhs.type
+            && lhs.url == rhs.url
+            && lhs.imageURL == rhs.imageURL
+            && lhs.file == rhs.file
+    }
+    
+    public func hash(into hasher: inout Hasher) {
+        hasher.combine(title)
+        hasher.combine(author)
+        hasher.combine(text)
+        hasher.combine(type)
+        hasher.combine(url)
+        hasher.combine(imageURL)
+        hasher.combine(file)
+    }
```

Comment:
I think it should be possible to let the compiler do this for us. `ExtraData` is `Hashable` too.

Reply:
Yeah, thanks. `AttachmentAction` was missing `Hashable` conformance.